src/SourceLocator/Type/AutoloadSourceLocator.php

```diff
@@ -163,18 +174,34 @@ private function locateClassByName(string $className) : ?string
 
         self::$autoloadLocatedFile = null;
         self::$currentAstLocator   = $this->astLocator; // passing the locator on to the implicitly instantiated `self`
-        $previousErrorHandler      = set_error_handler(static function (int $errno, string $errstr) : bool {
-            return true;
-        });
-        stream_wrapper_unregister('file');
-        stream_wrapper_register('file', self::class);
-        class_exists($className);
-        stream_wrapper_restore('file');
-        set_error_handler($previousErrorHandler);
+
+        $this->setErrorHandler();
+
+        foreach ($this->streamWrapperProtocols as $protocol) {
+            stream_wrapper_unregister($protocol);
+            stream_wrapper_register($protocol, self::class);
+        }
+
+        try {
```

Comment:
We should verify this `try-finally` by adding a throwing autoloader to the stack.

The test would be something like:

```php
$toBeThrown = new SpecificExceptionType();

spl_autoload_register(static function () use ($toBeThrown) {
    throw $toBeThrown;
});

try {
    // ... try to locate symbol
    self::fail('No exception was thrown');
} catch (SpecificExceptionType $caught) {
    self::assertSame($toBeThrown, $caught);
}

// verify error handler and stream wrapper here
```

Reply:
Errata: test already in place, just needs a couple assertions more :+1:

Reply:
I don't know how to test "verify error handler and stream wrapper here"

Both don't have "get" function.

For error handler I can do something like this:

```
        $previousErrorHandler = set_error_handler(function () {
            return true;
        });
        restore_error_handler();
```

However it's very ugly.

Reply:
The code for the "previous error handler" is exactly like that.

For the stream wrappers, maybe `stream_get_wrappers()`?

Reply:
It only returns the list of registered wrappers.

Reply:
Guess we can't test that detail then :shrug:

Reply:
Actually: we can! `file_get_contents(__FILE__)` should work fine if the stream wrapper is gone (it won't if we don't restore it)

Reply:
Added. However it looks that PHPUnit is broken without stream wrapper restore :)

```
Warning: include(P:\Git\better-reflection\vendor\phpunit\phpunit\src\Framework\Error\Warning.php): failed to open stream: "Roave\BetterReflection\SourceLocator\Type\AutoloadSourceLocator::stream_open" call failed
"Roave\BetterReflection\SourceLocator\Type\AutoloadSourceLocator::stream_open" call failed in P:\Git\better-reflection\vendor\composer\ClassLoader.php on line 444

Warning: include(): Failed opening 'P:\Git\better-reflection\vendor\composer/../phpunit/phpunit/src/Framework/Error/Warning.php' for inclusion (include_path='.;C:\php\pear') in P:\Git\better-reflection\vendor\composer\ClassLoader.php on line 444
```